chia/util/db_wrapper.py

```diff
@@ -29,3 +31,108 @@ async def rollback_transaction(self):
 
     async def commit_transaction(self) -> None:
         await self.db.commit()
+
+
+class DBWrapper2:
+    db_version: int
+    _lock: asyncio.Lock
+    _read_connections: asyncio.Queue[aiosqlite.Connection]
+    _write_connection: aiosqlite.Connection
+    _num_read_connections: int
+    _in_use: Dict[asyncio.Task, aiosqlite.Connection]
+    _current_writer: Optional[asyncio.Task]
+    _savepoint_name: int
+
+    async def add_connection(self, c: aiosqlite.Connection) -> None:
+        # this guarantees that reader connections can only be used for reading
+        await c.execute("pragma query_only")
+        self._read_connections.put_nowait(c)
+        self._num_read_connections += 1
+
+    def __init__(self, connection: aiosqlite.Connection, db_version: int = 1) -> None:
+        self._read_connections = asyncio.Queue()
+        self._write_connection = connection
+        self._lock = asyncio.Lock()
+        self.db_version = db_version
+        self._num_read_connections = 0
+        self._in_use = {}
+        self._current_writer = None
+        self._savepoint_name = 0
+
+    async def close(self) -> None:
+        while self._num_read_connections > 0:
+            await (await self._read_connections.get()).close()
+            self._num_read_connections -= 1
+        await self._write_connection.close()
+
+    def _next_savepoint(self) -> str:
+        name = f"s{self._savepoint_name}"
+        self._savepoint_name += 1
+        return name
+
+    @contextlib.asynccontextmanager
+    async def write_db(self) -> AsyncIterator[aiosqlite.Connection]:
+        task = asyncio.current_task()
+        assert task is not None
+        if self._current_writer == task:
+            # we allow nesting writers within the same task
```

Comment:
why do we allow this?

Reply:
(not allowing this could increase safety and code readability)

Reply:
we do have a few places where we call functions while also holding a write-enabled connection. I can look them up